File: src/DiagToolsValidator/DiagToolsValidator.Core/Functionality/DotNetFunctionality/DotNetApp.fs
namespace DiagToolsValidator.Core.Functionality

open System
open System.IO
open System.Xml.Linq
open System.Collections.Generic


module DotNetApp =
    type DotNetApp (dotNetEnv: Dictionary<string, string>,
                    appTemplate: string,
                    appRoot: string) =
        let _dotNetEnv = dotNetEnv
        let _appTemplate = appTemplate
        let _appRoot = appRoot

        member val AppRoot: string = _appRoot with get

        new() =
            let env = new Dictionary<string, string>()
            new DotNetApp(env, "", "")

        member x.GetProjectFile = Directory.GetFiles(_appRoot, "*.csproj")[0]
        
        member x.GetTargetFramework =
            try
                let projectFile = x.GetProjectFile
                let xmlData = File.ReadAllText(projectFile)
                let doc = XDocument.Parse(xmlData)
                Choice1Of2 (doc.Root.Element("PropertyGroup").Element("TargetFramework").Value)
            with
            | ex -> 
                ex.Data.Add("DotNetApp.GetTargetFramework", $"Fail to get target framework from {_appRoot}")
                Choice2Of2 ex

        member x.GetSymbolFolder (buildConfig: string) (targetRID: string) =
            let trace = new Core.ProgressTraceBuilder(null)
            trace {
                let! targetFramework = x.GetTargetFramework
                let appSymbolFolder = Path.Combine(_appRoot, "bin", buildConfig, targetFramework, targetRID)
                if Path.Exists(appSymbolFolder)
                then 
                    return appSymbolFolder
                else 
                    let ex = new Exception("Fail to get .NET app symbol folder")
                    ex.Data.Add("DotNetApp.GetAppSymbolFolder", $"Expected symbol folder {appSymbolFolder} doesn't exsit")
                    return! Choice2Of2 ex
            }

        member x.GetNativeSymbolFolder (buildConfig: string) (targetRID: string) =
            let trace = new Core.ProgressTraceBuilder(null)
            trace {
                let! targetFramework = x.GetTargetFramework
                let appSymbolFolder = Path.Combine(_appRoot, "bin", buildConfig, targetFramework, targetRID, "publish")
                if Path.Exists(appSymbolFolder)
                then 
                    return appSymbolFolder
                else 
                    let ex = new Exception("Fail to get .NET app native symbol folder")
                    ex.Data.Add("DotNetApp.GetNativeSymbolFolder", $"Expected native symbol folder {appSymbolFolder} doesn't exsit")
                    return! Choice2Of2 ex
            }

        member x.GetAppExecutable (buildConfig: string) (targetRID: string) =
            let trace = new Core.ProgressTraceBuilder(null)
            let excutableFileExtension = DotNet.GetExcutableFileExtensionByRID targetRID
            trace {
                let! symbolFolder = x.GetSymbolFolder buildConfig targetRID
                let projectFile = x.GetProjectFile
                let appName = Path.GetFileNameWithoutExtension(projectFile)
                let excutable = Path.Combine(symbolFolder, $"{appName}{excutableFileExtension}")
                if Path.Exists(excutable)
                then 
                    return excutable
                else 
                    let ex = new Exception("Fail to get .NET app excutable file")
                    ex.Data.Add("DotNetApp.GetAppExecutable", $"Excutable file {excutable} doesn't exsit")
                    return! Choice2Of2 ex
            }
            
        member x.GetAppNativeExecutable (buildConfig: string) (targetRID: string) =
            let trace = new Core.ProgressTraceBuilder(null)
            let excutableFileExtension = DotNet.GetExcutableFileExtensionByRID targetRID
            trace {
                let! symbolFolder = x.GetNativeSymbolFolder buildConfig targetRID
                let projectFile = x.GetProjectFile
                let appName = Path.GetFileNameWithoutExtension(projectFile)
                let excutable = Path.Combine(symbolFolder, $"{appName}{excutableFileExtension}")
                if Path.Exists(excutable)
                then 
                    return excutable
                else 
                    let ex = new Exception("Fail to get .NET app native excutable file")
                    ex.Data.Add("DotNetApp.GetAppNativeExecutable", $"Native excutable file {excutable} doesn't exsit")
                    return! Choice2Of2 ex
            }

        member x.GetCreateDump (buildConfig: string) (targetRID: string) =
            let trace = new Core.ProgressTraceBuilder(null)
            let excutableFileExtension = DotNet.GetExcutableFileExtensionByRID targetRID
            trace {
                let! symbolFolder = x.GetSymbolFolder buildConfig targetRID
                let createDump = Path.Combine(symbolFolder, $"createdump{excutableFileExtension}")
                if Path.Exists(createDump)
                then 
                    return createDump
                else 
                    let ex = new Exception("Fail to get createadump for .NET app")
                    ex.Data.Add("DotNetApp.GetCreateDump", $"{createDump} doesn't exsit")
                    return! Choice2Of2 ex
            }
            
        member x.CreateApp () =
            DotNet.RunDotNetCommand _dotNetEnv 
                                    $"new {_appTemplate} -o {_appRoot} --force"
                                    "" 
                                    true 
                                    CommandLineTool.PrinteOutputData 
                                    CommandLineTool.PrintErrorData 
                                    true

        member x.BuildApp (buildConfig: string) (targetRID: string) =
            DotNet.RunDotNetCommand _dotNetEnv
                                    $"build -r {targetRID} -c {buildConfig}"
                                    appRoot 
                                    true 
                                    CommandLineTool.PrinteOutputData 
                                    CommandLineTool.PrintErrorData 
                                    true
                                    
        member x.PublishApp (buildConfig: string) (targetRID: string) =
            DotNet.RunDotNetCommand _dotNetEnv
                                    $"publish -r {targetRID} -c {buildConfig}"
                                    appRoot 
                                    true 
                                    CommandLineTool.PrinteOutputData 
                                    CommandLineTool.PrintErrorData 
                                    true

File: src/DiagToolsTestRunner/CoreFunctionality/DotNetApp.py
import os
from xml.etree import ElementTree as ET

from CoreFunctionality.cli import CommandInvoker
from CoreFunctionality.DotNetEnvironment import DotNetEnvironment 

class DotNetApp:
    def __init__(self, dotnet_env: DotNetEnvironment, root: str, template: str, name: str=None):
        self.__root = root
        self.__template = template
        self.__dotnet_env = dotnet_env

        if name is None:
            self.__name = os.path.basename(root)
        else:
            self.__name = name

    def get_project_file_path(self):
        return os.path.join(self.__root, f'{self.__name}.csproj')
    
    def get_target_framework(self):
        project_file_path = self.get_project_file_path()
        tree = ET.parse(project_file_path)
        root = tree.getroot()
        return root.find('PropertyGroup').find('TargetFramework').text
    
    def get_symbol_folder(self, target_framework: str, build_config: str, target_rid: str, output_folder: str=None):
        if output_folder is None:
            output_folder = os.path.join(self.__root, 'bin')

        return os.path.join(output_folder, build_config, target_framework, target_rid)
    
    def get_executable(self, target_framework: str, build_config: str, target_rid: str, output_folder: str=None):
        symbol_folder = self.get_symbol_folder(target_framework, build_config, target_rid, output_folder)
        return os.path.join(
            symbol_folder,
            f'{self.__name}{DotNetEnvironment.get_executable_file_extension_by_rid(target_rid)}'
        )
    
    def get_native_symbol_folder(self, target_framework: str, build_config: str, target_rid: str, output_folder: str=None):
        if output_folder is None:
            output_folder = os.path.join(self.__root, 'bin')

        return os.path.join(output_folder, build_config, target_framework, target_rid, 'native')
    
    def get_native_executable(self, target_framework: str, build_config: str, target_rid: str, output_folder: str=None):
        native_symbol_folder = self.get_native_symbol_folder(target_framework, build_config, target_rid, output_folder)
        return os.path.join(
            native_symbol_folder,
            f'{self.__name}{DotNetEnvironment.get_executable_file_extension_by_rid(target_rid)}'
        )
    
    def create(self, redirect_std_out_err: bool=True, silent: bool=False):
        args = [
            self.__dotnet_env.dotnet_executable, 'new', self.__template,
            '-o', self.__root,
            '-n', self.__name,
            '--force'
        ]
        with CommandInvoker(
            args,
            env=self.__dotnet_env.environment_variables,
            redirect_std_out_err=redirect_std_out_err,
            silent=silent
        ) as p:
            p.communicate()
            return p
    
    def build(self, build_config: str, target_rid: str, redirect_std_out_err: bool=True, silent: bool=False):
        args = [
            self.__dotnet_env.dotnet_executable, 'build',
            '-r', target_rid,
            '-c', build_config
        ]
        with CommandInvoker(
            args,
            env=self.__dotnet_env.environment_variables,
            cwd=self.__root,
            redirect_std_out_err=redirect_std_out_err,
            silent=silent
        ) as p:
            p.communicate()
            return p
        
    def publish(self, build_config: str, target_rid: str, redirect_std_out_err: bool=True, silent: bool=False):
        args = [
            self.__dotnet_env.dotnet_executable, 'publish',
            '-r', target_rid,
            '-c', build_config
        ]
        with CommandInvoker(
            args,
            env=self.__dotnet_env.environment_variables,
            cwd=self.__root,
            redirect_std_out_err=redirect_std_out_err,
            silent=silent
        ) as p:
            p.communicate()
            return p